roachprod/cli: add disk-stall chaos command with cgroup support

shailendra-patel opened the pull request:
Previously, roachprod chaos commands only supported network failures (partitions and latency). There was no CLI support for injecting disk stalls to test CockroachDB's storage resilience.

This commit adds a new `roachprod chaos disk-stall` command that integrates with the existing failure injection library to inject disk I/O stalls using cgroups v2.

Example usage:
```shell
roachprod chaos disk-stall mycluster --type cgroup --nodes 1,2,3

roachprod chaos disk-stall mycluster --type cgroup --nodes 1 --insecure

#using dmsetup
roachprod chaos disk-stall mycluster --type dmsetup --nodes 1 --insecure
```

Epic: none

Release note: None

shailendra-patel committed:
roachprod/cli: add disk-stall chaos command with cgroup support

Previously, roachprod chaos commands only supported network failures
(partitions and latency). There was no CLI support for injecting disk
stalls to test CockroachDB's storage resilience.

This commit adds a new `roachprod chaos disk-stall` command that
integrates with the existing failure injection library to inject disk
I/O stalls using cgroups v2.

Example usage:
```
roachprod chaos disk-stall mycluster --type cgroup --nodes 1,2,3

roachprod chaos disk-stall mycluster --type cgroup --nodes 1 --insecure
```

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/158457)
<!-- Reviewable:end -->

shailendra-patel committed:
Add dmsetup as a new disk staller type that uses device-mapper to
completely freeze disk I/O, more closely simulating hardware failure
than cgroup throttling. The device name is persisted to a state file
during setup so cleanup works when running stages independently.

Add --stage=inject-recover option that runs inject, waits for the
configured duration, then recovers. This is useful for running
setup and cleanup as separate steps.

Make dmsetup WaitForFailureToPropagate a no-op since clusters with
WAL failover enabled continue operating during single-store disk stalls.

Epic: none

Release note: None

DarrylWong commented on pkg/roachprod/failureinjection/failures/disk_stall.go:
```diff
@@ -602,18 +648,13 @@ func (s *DmsetupDiskStaller) Cleanup(
-	return forEachNode(nodes, func(n install.Nodes) error {
-		return s.WaitForSQLUnavailable(ctx, l, n, 3*time.Minute)
-	})
+	// No-op: With WAL failover enabled, CRDB clusters continue operating normally
```
Hmm can we just skip calling WaitForFailureToPropagate/make it toggle-able? This would also apply to cgroup too right?

shailendra-patel replied:
Thanks for clarifying. I have introduced a DiskStall Arg to skip WaitForFailureToPropagate.

DarrylWong reviewed (commented):
One comment but otherwise seems good to me.

DarrylWong commented on pkg/cmd/roachprod/cli/chaos_disk.go:
```diff
@@ -0,0 +1,160 @@
+					Cycle:                  cycle,
+					CycleStallDuration:     cycleStallDuration,
+					CycleUnstallDuration:   cycleUnstallDuration,
+					SkipFailurePropagation: true,
```
Hmm, why can't we just not call it from the CLI's end? It's not required for correctness from the frameworks end.

shailendra-patel commented:
TFTR @DarrylWong 

bors r = DarrylWong
